### PR TITLE
Add doc for table_changes function

### PR DESCRIPTION
**Pull request by homar**
## Description
Docs changes for https://github.com/trinodb/trino/pull/16205

## Release notes

(x) This is not user-visible or docs only and no release notes are required.

### CODE REVIEW

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 1]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
 * ``$file_size``
     Size of the file for this row.
 
+Functions
+---------
+
+The connector exposes the following table functions:
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:728]:
```suggestion
The connector provides the following table functions:
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 2]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
...
+
+system.table_changes
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:730]:
```suggestion
table_changes
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 3]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
...
+^^^^^^^^^^^^^^^^^^^^
+
+Allows reading Change Data Feed (CDF) entries to
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:733]: Wrapping is still weird for this paragraph .. 80 char would be ideal

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 4]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
...
+expose row-level changes between two versions of a Delta Lake table.
+When the ``change_data_feed_enabled`` table property is set to ``true``
+on a specific Delta Lake table, the connector records change events for
+all data changes on the table.
+In addition to columns present in the table this function
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:738]:
```suggestion
In addition to columns present in the table
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 5]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
...
+the function returns the following values for each change event:
+
+* ``_change_type``
+    Type ``VARCHAR``, gives the type of change that occurred,
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:742]:
```suggestion
    Gives the type of change that occurred as represented as ``VARCHAR`` value. 
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 6]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
...
+    can be one of the following:``insert``, ``delete``, ``update_preimage``, ``update_postimage``
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:743]:
```suggestion
    Possible values are ``insert``, ``delete``, ``update_preimage``, and ``update_postimage``.
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 7]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
...
+
+* ``_commit_version``
+    Type ``BIGINT``, shows the table version for which
+    the change occurred
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:747]:
```suggestion

```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 8]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
...
+
+* ``_commit_timestamp``
+    Type ``TIMESTAMP(3) WITH TIME ZONE``,
+    shows the timestamp of the commit in which the specified change happened
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:751]:
```suggestion
    Represents the timestamp for the commit in which the specified change happened as 
    ``TIMESTAMP(3) WITH TIME ZONE`` data type value.
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 9]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
...
+.. code-block:: sql
+
+    CREATE TABLE pages (page_url VARCHAR, domain VARCHAR, views INTEGER)
+        WITH (change_data_feed_enabled = true);
+
+Insert some data:
+
+.. code-block:: sql
+
+    INSERT INTO pages VALUES
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:781]: I think we should at least move values down into a separate line

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 10]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
...
+        ('url4', 'domain1', 400),
+        ('url5', 'domain2', 500),
+        ('url6', 'domain3', 2);
+
+Update some data:
+
+.. code-block:: sql
+
+    UPDATE pages SET domain = 'domain4' WHERE views = 2;
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:794]: three lines

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 11]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
...
+  page_url    |     domain     |    views    |    _change_type     |    _commit_version    |    _commit_timestamp
+  url4        |     domain1    |    400      |    insert           |     2                 |    2023-03-10T21:22:23.000+0000
+  url5        |     domain2    |    500      |    insert           |     2                 |    2023-03-10T21:22:23.000+0000
+  url6        |     domain3    |    2        |    insert           |     2                 |    2023-03-10T21:22:23.000+0000
+  url2        |     domain2    |    2        |    update_preimage  |     3                 |    2023-03-10T22:23:24.000+0000
+  url2        |     domain4    |    2        |    update_postimage |     3                 |    2023-03-10T22:23:24.000+0000
+  url6        |     domain3    |    2        |    update_preimage  |     3                 |    2023-03-10T22:23:24.000+0000
+  url6        |     domain4    |    2        |    update_postimage |     3                 |    2023-03-10T22:23:24.000+0000
+
+In the output it can be easily seen what changes happen in which version.
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:825]:
```suggestion
The output shows what changes happen in which version.
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 12]
```diff
@@ -722,6 +722,176 @@ directly or used in conditional statements.
...
+For example in version 3 two rows were modified, first one changed from
+``('url2', 'domain2', 2)`` into ``('url2', 'domain4', 2)`` and the second from
+``('url6', 'domain2', 2)`` into ``('url6', 'domain4', 2)``.
+
+If ``since_version`` is not provided the function produces change events
+starting from when the table was created.
+
+Example:
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:833]: From 833 to 861 .. this is all the same as above right. .. if so .. just omit it all

**Review by mosabua — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/connector/delta-lake.rst (hunk 5 above)
Comment (homar) [reply docs/src/main/sphinx/connector/delta-lake.rst:742]: I will change it but tbh all three descriptions of these additional columns that you propose seems unnaturally long and weird to me. I tried to be as short and as precise as possible. https://docs.databricks.com/delta/delta-change-data-feed.html#what-is-the-schema-for-the-change-data-feed - this is how databricks describes them

File: docs/src/main/sphinx/connector/delta-lake.rst (hunk 10 above)
Comment (homar) [reply docs/src/main/sphinx/connector/delta-lake.rst:794]: sorry i don't get this comment :|

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 13]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+---------------
+
+The connector provides the following table functions:
+
+table_changes
+^^^^^^^^^^^^^
+
+Allows reading Change Data Feed (CDF) entries to expose row-level changes
+between two versions of a Delta Lake table.
+When the ``change_data_feed_enabled`` table property is set to ``true``
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:735]: Wrap up

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 14]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+on a specific Delta Lake table, the connector records change events for
+all data changes on the table. In addition to columns present in the table
+the function returns the following values for each change event:
+
+* ``_change_type``
+    Gives the type of change that occurred as represented as ``VARCHAR`` value.
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:741]:
```suggestion
    Gives the type of change that occurred.
```

and wrap

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 15]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+    Possible values are ``insert``, ``delete``, ``update_preimage`` and ``update_postimage``.
+
+* ``_commit_version``
+    Shows the table version for which the change occurred as ``BIGINT`` value.
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:745]:
```suggestion
    Shows the table version for which the change occurred.
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 16]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+
+* ``_commit_timestamp``
+    Represents the timestamp for the commit in which the specified change happened as
+    ``TIMESTAMP(3) WITH TIME ZONE`` data type value.
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:749]:
```suggestion
   
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 17]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+
+* ``_change_type``
+    Gives the type of change that occurred as represented as ``VARCHAR`` value.
+    Possible values are ``insert``, ``delete``, ``update_preimage`` and ``update_postimage``.
+
+* ``_commit_version``
+    Shows the table version for which the change occurred as ``BIGINT`` value.
+
+* ``_commit_timestamp``
+    Represents the timestamp for the commit in which the specified change happened as
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:748]:
```suggestion
    Represents the timestamp for the commit in which the specified change happened.
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 18]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+    FROM
+      TABLE(
+        system.table_changes(
+          schema_name => 'test_schema',
+          table_name => 'tableName',
+          since_version => 0
+        )
+      );
+
+``schema_name`` - type ``VARCHAR``, required, name of the schema for which the function will be called
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:764]:
```suggestion
``schema_name`` - type ``VARCHAR``, required, name of the schema for which the function is called
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 19]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+
+``table_name`` - type ``VARCHAR``, required, name of the table for which the function will be called
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:766]:
```suggestion
``table_name`` - type ``VARCHAR``, required, name of the table for which the function is called
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 20]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+
+``since_version`` - type ``BIGINT``, optional, version from which changes will be shown, exclusive
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:768]:
```suggestion
``since_version`` - type ``BIGINT``, optional, version from which changes are shown, exclusive
```

Also what do ou mean by exclusive? If you mean that it work show that version itself we should maybe reword

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 21]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+Example:
+
+Create table:
+
+.. code-block:: sql
+
+    CREATE TABLE pages (page_url VARCHAR, domain VARCHAR, views INTEGER)
+        WITH (change_data_feed_enabled = true);
+
+Insert some data:
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:779]:
```suggestion
Insert data:
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 22]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+            ('url1', 'domain1', 1),
+            ('url2', 'domain2', 2),
+            ('url3', 'domain1', 3);
+    INSERT INTO pages
+        VALUES
+            ('url4', 'domain1', 400),
+            ('url5', 'domain2', 500),
+            ('url6', 'domain3', 2);
+
+Update some data:
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:794]:
```suggestion
Update data:
```

File: docs/src/main/sphinx/connector/delta-lake.rst (hunk 10 above)
Comment (mosabua) [reply docs/src/main/sphinx/connector/delta-lake.rst:798]:
```
 UPDATE pages 
   SET domain = 'domain4' 
   WHERE views = 2;
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 23]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+Select changes:
+
+.. code-block:: sql
+
+    SELECT
+      *
+    FROM
+      TABLE(
+        system.table_changes(
+          schema_name => 'test_schema',
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:809]: I think we need to add the schema to the queries above to be clear

Comment (mosabua) [reply docs/src/main/sphinx/connector/delta-lake.rst:809]: Or add a "use" command first

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 24]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+
+The output shows what changes happen in which version.
+For example in version 3 two rows were modified, first one changed from
+``('url2', 'domain2', 2)`` into ``('url2', 'domain4', 2)`` and the second from
+``('url6', 'domain2', 2)`` into ``('url6', 'domain4', 2)``.
+
+If ``since_version`` is not provided the function produces change events
+starting from when the table was created.
+
+Example, provided that there is exactly the same table as created above:
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:837]: We can omit that .. its implied since we are still talking about the same example

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 25]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+    FROM
+      TABLE(
+        system.table_changes(
+          schema_name => 'test_schema',
+          table_name => 'pages'
+        )
+      )
+    ORDER BY _commit_version ASC;
+
+The preceding sequence of SQL statements returns the following result:
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:852]:
```suggestion
The preceding SQL statement returns the following result:
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 26]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+  url3        |     domain1    |    3        |    insert           |     1                 |    2023-03-10T20:21:22.000+0000
+  url4        |     domain1    |    400      |    insert           |     2                 |    2023-03-10T21:22:23.000+0000
+  url5        |     domain2    |    500      |    insert           |     2                 |    2023-03-10T21:22:23.000+0000
+  url6        |     domain3    |    2        |    insert           |     2                 |    2023-03-10T21:22:23.000+0000
+  url2        |     domain2    |    2        |    update_preimage  |     3                 |    2023-03-10T22:23:24.000+0000
+  url2        |     domain4    |    2        |    update_postimage |     3                 |    2023-03-10T22:23:24.000+0000
+  url6        |     domain3    |    2        |    update_preimage  |     3                 |    2023-03-10T22:23:24.000+0000
+  url6        |     domain4    |    2        |    update_postimage |     3                 |    2023-03-10T22:23:24.000+0000
+
+Here we can see changes that occurred at version 1 - three inserts - they were
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:868]:
```suggestion
You can see changes that occurred at version 1 as- three inserts. They are
```

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 27]
```diff
@@ -722,6 +722,152 @@ directly or used in conditional statements.
...
+not visible in the previous code snippet when since_version value was set to 1.
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:869]:
```suggestion
not visible in the previous statement when since_version value was set to 1.
```

**Review by mosabua — CHANGES_REQUESTED** [review]

**Comment by mosabua** [comment]
I simplified a bit..

**Comment by mosabua** [comment]
Let me know when you have the code PR ready and this updates so I can review for latest status. Also currently this PR conflicts..

**Comment by mosabua** [comment]
@findepi merged the code PR. @homar can you please make sure you resolve the conflicts and address any comments as soon as possible so we can make sure we get this PR merged before the release next week. Also adapt the docs to whatever changed in the code before merge ..

**Comment by homar** [comment]
> @findepi merged the code PR. @homar can you please make sure you resolve the conflicts and address any comments as soon as possible so we can make sure we get this PR merged before the release next week. Also adapt the docs to whatever changed in the code before merge ..

Yeah sorry, I was in warsaw office for 2 days and I didn't have time to update this one. I have updated it only now

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 28]
```diff
@@ -732,6 +732,152 @@ Fault-tolerant execution support
...
+The connector provides the following table functions:
+
+table_changes
+^^^^^^^^^^^^^
+
+Allows reading Change Data Feed (CDF) entries to expose row-level changes
+between two versions of a Delta Lake table. When the ``change_data_feed_enabled``
+table property is set to ``true`` on a specific Delta Lake table,
+the connector records change events for all data changes on the table.
+In addition to columns present in the table the function returns the following
```
Comment (colebow) [inline docs/src/main/sphinx/connector/delta-lake.rst:748]:
```suggestion
In addition to returning the columns present in the table, the function
returns the following
```

**Review by colebow — APPROVED** [review]
One nit, but LGTM.

File: docs/src/main/sphinx/connector/delta-lake.rst (hunk 28 above)
Comment (mosabua) [reply docs/src/main/sphinx/connector/delta-lake.rst:748]: Please apply this suggestion @homar

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 29]
```diff
@@ -732,6 +732,152 @@ Fault-tolerant execution support
...
+    Gives the type of change that occurred. Possible values are ``insert``,
+    ``delete``, ``update_preimage`` and ``update_postimage``.
+
+* ``_commit_version``
+    Shows the table version for which the change occurred.
+
+* ``_commit_timestamp``
+    Represents the timestamp for the commit in which the specified change happened.
+
+.. code-block:: sql
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:761]: This shows how to run the table function.. and I think it should therefore be explained before preceding info that explains the returned data

We also need a entry sentence before the query

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 30]
```diff
@@ -732,6 +732,152 @@ Fault-tolerant execution support
...
+        )
+      );
+
+``schema_name`` - type ``VARCHAR``, required, name of the schema for which the function is called
+
+``table_name`` - type ``VARCHAR``, required, name of the table for which the function is called
+
+``since_version`` - type ``BIGINT``, optional, version from which changes are shown, exclusive
+
+Example:
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:780]: Expand this to a sentence.

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 31]
```diff
@@ -732,6 +732,152 @@ Fault-tolerant execution support
...
+  url3        |     domain1    |    3        |    insert           |     1                 |    2023-03-10T20:21:22.000+0000
+  url4        |     domain1    |    400      |    insert           |     2                 |    2023-03-10T21:22:23.000+0000
+  url5        |     domain2    |    500      |    insert           |     2                 |    2023-03-10T21:22:23.000+0000
+  url6        |     domain3    |    2        |    insert           |     2                 |    2023-03-10T21:22:23.000+0000
+  url2        |     domain2    |    2        |    update_preimage  |     3                 |    2023-03-10T22:23:24.000+0000
+  url2        |     domain4    |    2        |    update_postimage |     3                 |    2023-03-10T22:23:24.000+0000
+  url6        |     domain3    |    2        |    update_preimage  |     3                 |    2023-03-10T22:23:24.000+0000
+  url6        |     domain4    |    2        |    update_postimage |     3                 |    2023-03-10T22:23:24.000+0000
+
+You can see changes that occurred at version 1 as- three inserts. They are
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:878]: as- ??

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 32]
```diff
@@ -732,6 +732,152 @@ Fault-tolerant execution support
...
+not visible in the previous statement when since_version value was set to 1.
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/delta-lake.rst:879]:
```suggestion
not visible in the previous statement when ``since_version`` value was set to 1.
```

**Review by mosabua — CHANGES_REQUESTED** [review]
Needs some reordering

**Comment by mosabua** [comment]
@homar .. are you updating this PR so we can review again and maybe merge soon?

**Review by mosabua — APPROVED** [review]
Looks good now. Let's ship it.

**Review by mosabua — APPROVED** [review]
Looks good now. Let's ship it.
